homeassistant/components/sensor/history_stats.py

```diff
@@ -0,0 +1,303 @@
+    vol.Required(CONF_STATE): cv.slug,
+    vol.Optional(CONF_START, default=None): cv.template,
+    vol.Optional(CONF_END, default=None): cv.template,
+    vol.Optional(CONF_DURATION, default=None): cv.template,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+})
+
+
+# noinspection PyUnusedLocal
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up the History Stats sensor."""
+    entity_id = config.get(CONF_ENTITY_ID)
+    entity_state = config.get(CONF_STATE)
+    start = config.get(CONF_START)
+    end = config.get(CONF_END)
+    duration = config.get(CONF_DURATION)
+    name = config.get(CONF_NAME)
+    hsh = HistoryStatsHelper
+
+    params = 0
+    for param in [start, end, duration]:
+        if param is not None:
+            params += 1
+
+    if params != 2:
+        raise _LOGGER.error('You must provide exactly 2 of the following:' +
+                            ' start, end, duration')
+
+    for template in [start, end, duration]:
+        if template is not None:
+            template.hass = hass
+            template.template = hsh.parse_time_expr(template.template)
+            # pylint: disable=protected-access
+            template._compiled_code = None  # Force recompilation
+
+    yield from async_add_devices([HistoryStatsSensor(
+        hass, entity_id, entity_state, start, end, duration, name)], True)
+    return True
+
+
+class HistoryStatsSensor(Entity):
+    """Representation of a HistoryStats sensor."""
+
+    def __init__(
+            self, hass, entity_id, entity_state, start, end, duration, name):
+        """Initialize the HistoryStats sensor."""
+        self._hass = hass
+
+        self._entity_id = entity_id
+        self._entity_state = entity_state
+        self._duration = duration
+        self._start = start
+        self._end = end
+        self._name = name
+        self._unit_of_measurement = UNIT
+
+        self._period = (0.0, 0.0)
+        self.value = 0
+
+        # noinspection PyUnusedLocal
+        # pylint: disable=invalid-name
+        @callback
+        def async_stats_sensor_state_listener(entity, old_state, new_state):
+            """Called when the sensor changes state."""
+            hass.async_add_job(self.async_update_ha_state, True)
+
+        async_track_state_change(
+            hass, entity_id, async_stats_sensor_state_listener)
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return round(self.value, 2)
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit the value is expressed in."""
+        return self._unit_of_measurement
+
+    @property
+    def should_poll(self):
+        """Polling required."""
+        return True
+
+    @property
+    def state_attributes(self):
```

Comment:
I'll change that, but just to let you know, the method is called `state_attributes` in `sensor/statisctics.py`. 

I used it to help me code my component.